### PR TITLE
Add SimplSchema validation logic to insert methods

### PR DESCRIPTION
**Pull request by MarvinZeising**

### CODE REVIEW

File: imports/api/notes/service.js [hunk 1]
```diff
@@ -68,7 +70,34 @@ Meteor.methods({
...
+    }).newContext()
+
+    validationContext.validate({ projectId, ...note })
+
+    if (!validationContext.isValid()) {
+      throw new ValidationError(validationContext.validationErrors())
+    }
+
+    if (!Meteor.userId()) {
+      throw Meteor.Error('must be logged in to insert a new project')
```
Comment (muhkuhxy) [inline imports/api/notes/service.js]: project != note

File: imports/api/projects/service.js [hunk 2]
```diff
@@ -73,7 +73,7 @@ Meteor.methods({
     validationContext.validate(project)
 
     if (!validationContext.isValid()) {
-      throw new Meteor.Error(new ValidationError(validationContext.validationErrors()))
+      throw new ValidationError(validationContext.validationErrors())
```
Comment (muhkuhxy) [inline imports/api/projects/service.js]: ehhh, this breaks the InsertForm now

File: imports/api/notes/service.js [hunk 3]
```diff
@@ -68,7 +70,34 @@ Meteor.methods({
...
+        }
+      },
+      title: String,
+      text: String
+    }).newContext()
+
+    validationContext.validate({ projectId, ...note })
+
+    if (!validationContext.isValid()) {
+      throw new ValidationError(validationContext.validationErrors())
```
Comment (muhkuhxy) [inline imports/api/notes/service.js]: should be new Meteor.Error(new ValidationError(...)) to work with InsertForm

File: imports/api/vessels/service.js [hunk 4]
```diff
@@ -62,7 +65,58 @@ Meteor.methods({
...
+      mmsi: {
+        type: String,
+        optional: true
+      }
+    }).newContext()
+
+    validationContext.validate({ projectId, ...vessel })
+
+    if (!validationContext.isValid()) {
+      throw new ValidationError(validationContext.validationErrors())
```
Comment (muhkuhxy) [inline imports/api/vessels/service.js]: same as above

File: imports/api/notes/service.js [hunk 5]
```diff
@@ -68,7 +70,34 @@ Meteor.methods({
     return getExtendedNote(projectId, noteId)[key]
   },
   'note.insert' ({ projectId }, note) {
-    checkPermissions(projectId)
+    const validationContext = new SimpleSchema({
+      projectId: {
+        type: String,
+        custom () {
+          const project = Projects.findOne(this.value, { fields: { _id: 1 } })
```
Comment (muhkuhxy) [inline imports/api/notes/service.js]: Not sure about this validation as part of the schema. Until now I thought of the schema as something similar to the types in a static typed language. On the other hand, this way enables us to extract common stuff like projectId validation into a common method... Until it's forgotten on a new method.
Validation in the method body has a higher visibility, I feel like.

File: imports/api/vessels/service.js [hunk 6]
```diff
@@ -90,7 +144,49 @@ Meteor.methods({
...
+    }).newContext()
+
+    validationContext.validate({ projectId, vesselId, ...visit })
+
+    if (!validationContext.isValid()) {
+      throw new ValidationError(validationContext.validationErrors())
+    }
+
+    if (!Meteor.userId()) {
+      throw Meteor.Error('must be logged in to insert a new project')
```
Comment (muhkuhxy) [inline imports/api/vessels/service.js]: project != vessel

File: imports/api/vessels/service.js [hunk 7]
```diff
@@ -187,7 +283,54 @@ Meteor.methods({
     }
   },
   'vessel.visit.language.insert' ({ projectId, vesselId, visitId }, { languageIds }) {
-    checkVesselModule(projectId)
+    const validationContext = new SimpleSchema({
+      projectId: {
+        type: String,
+        custom () {
+          const project = Projects.findOne(this.value, { fields: { vesselModule: 1 } })
```
Comment (muhkuhxy) [inline imports/api/vessels/service.js]: If it stays this way, I think this merits extracting a projectId validation method.

File: imports/api/vessels/service.js [hunk 8]
```diff
@@ -187,7 +283,54 @@ Meteor.methods({
...
+
+          if (!project || !project.vesselModule) {
```
Comment (muhkuhxy) [inline imports/api/vessels/service.js]: In a common method, additional validations like `!project.vesselModule` could be handled by a lambda parameter

File: imports/api/vessels/service.js [hunk 9]
```diff
@@ -187,7 +283,54 @@ Meteor.methods({
...
+    }).newContext()
+
+    validationContext.validate({ projectId, vesselId, visitId, languageIds })
+
+    if (!validationContext.isValid()) {
+      throw new ValidationError(validationContext.validationErrors())
+    }
+
+    if (!Meteor.userId()) {
+      throw Meteor.Error('must be logged in to insert a new project')
```
Comment (muhkuhxy) [inline imports/api/vessels/service.js]: visit != project

**Comment by muhkuhxy** [comment]
This is merging onto branch `newProject`, not develop. Is that on purpose?

File: imports/api/notes/service.js (hunk 3 above)
Comment (MarvinZeising) [reply imports/api/notes/service.js]: Yes, that's exactly what I'm trying to do :-)

File: imports/api/vessels/service.js (hunk 7 above)
Comment (MarvinZeising) [reply imports/api/vessels/service.js]: Good point

File: imports/api/notes/service.js (hunk 5 above)
Comment (MarvinZeising) [reply imports/api/notes/service.js]: Well, it's one option to do it. I guess you have a point with the higher visibility.

**Comment by MarvinZeising** [comment]
Yep, that was on purpose. But merging it into develop seems fine, now, as well.

File: imports/api/projects/service.js (hunk 2 above)
Comment (MarvinZeising) [reply imports/api/projects/service.js]: Right, I should re-add your changes for the InsertForm

File: imports/api/notes/service.js (hunk 3 above)
Comment (muhkuhxy) [reply imports/api/notes/service.js]: But is that compatible with the errors coming from PersistenceManager.validate?

**Pull request merged**